types/src/v0/impls/l1.rs

```diff
@@ -870,51 +870,177 @@ impl L1Client {
             .await
     }
 
-    /// Get `StakeTable` at block height.
+    /// Get `StakeTable` at specific l1 block height.
+    /// This function fetches and processes various events (ValidatorRegistered, ValidatorExit,
+    /// Delegated, Undelegated, and ConsensusKeysUpdated) within the block range from the
+    /// contract's initialization block or `0`` is not provided to the provided `to_block` value.
+    /// Events are fetched in chunks to and retries are implemented for failed requests.
     pub async fn get_stake_table(
         &self,
         contract: Address,
-        block: u64,
+        to_block: u64,
     ) -> anyhow::Result<IndexMap<Address, Validator<BLSPubKey>>> {
-        // TODO stake_table_address needs to be passed in to L1Client
-        // before update loop starts.
         let stake_table_contract = StakeTable::new(contract, self.provider.clone());
 
-        let registered = stake_table_contract
-            .ValidatorRegistered_filter()
-            .from_block(0)
-            .to_block(block)
-            .query()
-            .await?;
+        // Retrieve the L1 block number when this contract was initialized.
+        // This helps avoid starting the event fetching from block number 0.
+        let from = if let Ok(init_block) = stake_table_contract.initializedAtBlock().call().await {
```

Comment:
@imabdulbasit we should retry this too

Reply:
Yes it gets set when the contract is deployed.

Reply:
https://github.com/EspressoSystems/espresso-network/pull/2985/commits/ca16b3df4164d7d849af656d2fce5422ccdacc77